### PR TITLE
New admin stock management

### PR DESCRIPTION
**Pull request by richardnuno**
Adds the ability to manage stock through a dedicated admin UI. The new UI allows an admin to make updates without incurring a full page reload and without having to select the variant to update in a separate form.

It also introduces a variant search search form that uses Spree::Core::Search::Variant to filter the returned results. The same UI is present in the product's stock management section but the
variant search results are scoped to the current product.

The Configuration admin tab was renamed to Settings to allow more visual space for the newly added Stock tab.

### CODE REVIEW

File: api/app/controllers/spree/api/stock_items_controller.rb [hunk 1]
```diff
@@ -57,9 +56,17 @@ def destroy
...
-      def stock_location
+      def load_readable_stock_locations
+        load_stock_location_by_ability(:read)
+      end
+
+      def load_updateable_stock_locations
+        load_stock_location_by_ability(:update)
+      end
+
+      def load_stock_location_by_ability(ability)
```
Comment (athal7) [inline api/app/controllers/spree/api/stock_items_controller.rb]: i think this parameter is commonly referred to as 'action'. i'm also not sure this level of abstraction is getting us all that much, could just put it in the controller action.

File: api/app/controllers/spree/api/stock_items_controller.rb [hunk 2]
```diff
@@ -33,16 +34,14 @@ def create
...
         @stock_item = StockItem.accessible_by(current_ability, :update).find(params[:id])
 
-        count_on_hand = 0
+        count_on_hand_adjustment = 0
         if params[:stock_item].has_key?(:count_on_hand)
-          count_on_hand = params[:stock_item][:count_on_hand].to_i
+          count_on_hand_adjustment = params[:stock_item][:count_on_hand].to_i
           params[:stock_item].delete(:count_on_hand)
         end
+        count_on_hand_adjustment -= @stock_item.count_on_hand if params[:stock_item][:force]
```
Comment (athal7) [inline api/app/controllers/spree/api/stock_items_controller.rb]: i'm not sure this is necessary with the new design, aren't we always setting the exact number, and can therefore just use `set_count_on_hand` always?

File: api/app/controllers/spree/api/stock_items_controller.rb [hunk 3]
```diff
@@ -70,6 +77,14 @@ def scope
       def stock_item_params
         params.require(:stock_item).permit(permitted_stock_item_attributes)
       end
+
+      def adjust_stock_item_count_on_hand(count_on_hand_adjustment)
+        if @stock_item.should_track_inventory?
```
Comment (athal7) [inline api/app/controllers/spree/api/stock_items_controller.rb]: i like that this makes us not create stock movements in this use case, but there are still other code paths that do create stock movements when track inventory is off, so that could create even more confusion. would be great to pull this logic out of the controller to somewhere that gets reused anytime we want to do a stock movement, rather than just making the good behavior here.

File: api/app/controllers/spree/api/stock_items_controller.rb (hunk 1 above)
Comment (richardnuno) [inline api/app/controllers/spree/api/stock_items_controller.rb]: I agree on the parameter rename but I think the shared method is useful since we want to return the 422 in both cases if the stock_location_id parameters is not present.

File: api/app/controllers/spree/api/stock_items_controller.rb (hunk 2 above)
Comment (richardnuno) [inline api/app/controllers/spree/api/stock_items_controller.rb]: I wanted to leave the ability to either specify the quantity or the adjustment in place since it's the api. Happy to remove it if no one thinks it'll be useful.

File: backend/app/controllers/spree/admin/concerns/stock_management.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      module StockManagement
+        extend ActiveSupport::Concern
+
+        @@variant_display_attributes = [
+          { string_key: :sku, attr_name: :sku },
+          { string_key: :name, attr_name: :name }
+        ]
+
+        def self.set_variant_display_attributes(display_attributes)
+          @@variant_display_attributes = display_attributes
```
Comment (athal7) [inline backend/app/controllers/spree/admin/concerns/stock_management.rb]: if you set @@ on a module method, does that end up setting the variable on the module's parent? i.e. `Module` ?

File: backend/app/controllers/spree/admin/products_controller.rb [hunk 5]
```diff
@@ -67,9 +68,10 @@ def clone
       end
 
       def stock
```
Comment (athal7) [inline backend/app/controllers/spree/admin/products_controller.rb]: maybe we just have this route go to the StockItemsController rather than having to include a module and support a non-restful route?

File: backend/app/controllers/spree/admin/stock_items_controller.rb [hunk 6]
```diff
@@ -33,6 +36,10 @@ def stock_movement_params
         def determine_backorderable
           @stock_item.backorderable = params[:stock_item].present? && params[:stock_item][:backorderable].present?
         end
+
+        def variant_scope
+          Spree::Variant.all
```
Comment (athal7) [inline backend/app/controllers/spree/admin/stock_items_controller.rb]: no accessibility check?

File: backend/app/assets/javascripts/spree/backend/stock_management.js.coffee [hunk 7]
```diff
@@ -1,9 +1,160 @@
-jQuery ->
-  $('.stock_item_backorderable').on 'click', ->
-    $(@).parent('form').submit()
-  $('.toggle_stock_item_backorderable').on 'submit', ->
+$(document).ready ->
```
Comment (athal7) [inline backend/app/assets/javascripts/spree/backend/stock_management.js.coffee]: would be great to a) not set all of these functions on the window, and b) not set all of these event handlers for all pages  given that they are only useful on one page.

File: backend/app/controllers/spree/admin/concerns/stock_management.rb [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+          @stock_item_stock_locations = selected_stock_location.present? ? [selected_stock_location] : @stock_locations
+        end
+
+        def find_selected_stock_location
+          @stock_locations.find { |sl| sl.id == params[:stock_location_id].to_i }
+        end
+
+        def load_filtered_variants
+          results = if params[:variant_search_term].blank?
+            variant_scope
```
Comment (athal7) [inline backend/app/controllers/spree/admin/concerns/stock_management.rb]: this seems like valuable logic to live in the searcher class, i would imagine it being reused.

File: api/app/controllers/spree/api/stock_items_controller.rb (hunk 2 above)
Comment (athal7) [inline api/app/controllers/spree/api/stock_items_controller.rb]: ah no that is good to keep the api consistent, thanks for the reminder

File: backend/app/controllers/spree/admin/stock_items_controller.rb (hunk 6 above)
Comment (richardnuno) [inline backend/app/controllers/spree/admin/stock_items_controller.rb]: Good catch, adding

File: backend/app/assets/javascripts/spree/backend/stock_management.js.coffee (hunk 7 above)
Comment (richardnuno) [inline backend/app/assets/javascripts/spree/backend/stock_management.js.coffee]: These functions won't be attached to the window since they're within the document ready callback. The event handlers will only be attached on pages that have the `#listing_product_stock` DOM element: `return unless $('#listing_product_stock').length > 0`

File: backend/app/assets/stylesheets/spree/backend/sections/_products.scss [hunk 9]
```diff
@@ -30,88 +30,84 @@
 }
 
 #listing_product_stock {
```
Comment (athal7) [inline backend/app/assets/stylesheets/spree/backend/sections/_products.scss]: do we need any of these styles anymore?

Comment (athal7) [inline backend/app/assets/stylesheets/spree/backend/sections/_products.scss]: er these are all of the stock styles aren't they

File: backend/app/controllers/spree/admin/stock_items_controller.rb [hunk 10]
```diff
@@ -33,6 +40,24 @@ def stock_movement_params
...
+        def load_product
+          @product = Product.accessible_by(current_ability, :read).friendly.find(params[:product_slug]) if params[:product_slug]
+        end
+
+        def load_stock_management_data
+          @stock_locations = Spree::StockLocation.accessible_by(current_ability, :read)
+          @stock_item_stock_locations = params[:stock_location_id].present? ? @stock_locations.where(id: params[:stock_location_id]) : @stock_locations
+          @variant_display_attributes = self.class.variant_display_attributes
+          variants_results = Spree::Core::Search::Variant.new(params[:variant_search_term], scope: variant_scope).results
+          @variants = variants_results.order(id: :desc).page(params[:page]).per(params[:per_page] || Spree::Config[:orders_per_page])
```
Comment (athal7) [inline backend/app/controllers/spree/admin/stock_items_controller.rb]: i wonder if we should make this be `collection` or override `collection` to be an empty array, otherwise we are making an unnecessary call to fetch all stock items.

File: backend/app/views/spree/admin/products/stock.html.erb [hunk 11]
```diff
@@ -6,102 +6,22 @@
 <%= render :partial => 'spree/admin/shared/product_tabs', :locals => { :current => 'Stock Management' } %>
```
Comment (athal7) [inline backend/app/views/spree/admin/products/stock.html.erb]: this entire file can go away i think

File: backend/app/assets/stylesheets/spree/backend/sections/_products.scss (hunk 9 above)
Comment (richardnuno) [inline backend/app/assets/stylesheets/spree/backend/sections/_products.scss]: Yup, moving this to a dedicated file

File: backend/app/views/spree/admin/products/stock.html.erb (hunk 11 above)
Comment (richardnuno) [inline backend/app/views/spree/admin/products/stock.html.erb]: Good catch, will remove

File: backend/app/controllers/spree/admin/stock_items_controller.rb (hunk 10 above)
Comment (richardnuno) [inline backend/app/controllers/spree/admin/stock_items_controller.rb]: Good catch, I think in this case it makes sense to override `collection` to be an empty array since the "resource" is a stock item.

**Comment by richardnuno** [comment]
@athal7 latest commits address your PR comments

**Comment by richardnuno** [comment]
Also, rebased on top of current 2-2-dev

File: backend/app/assets/javascripts/spree/backend/stock_management.js.coffee (hunk 7 above)
Comment (athal7) [inline backend/app/assets/javascripts/spree/backend/stock_management.js.coffee]: wonderful

**Comment by athal7** [comment]
:shipit:

File: api/app/controllers/spree/api/stock_items_controller.rb [hunk 12]
```diff
@@ -16,14 +16,9 @@ def show
...
-        count_on_hand = 0
-        if params[:stock_item].has_key?(:count_on_hand)
-          count_on_hand = params[:stock_item][:count_on_hand].to_i
-        end
-
         @stock_item = scope.new(stock_item_params)
-        if @stock_item.save
-          @stock_item.adjust_count_on_hand(count_on_hand)
+
+        if @stock_item.save && adjust_stock_item_count_on_hand(count_on_hand_adjustment)
```
Comment (gmacdougall) [inline api/app/controllers/spree/api/stock_items_controller.rb]: Should these operations be wrapped in a transaction?

What happens when the stock item saves successfully and the adjust doesn't work for some reason?

File: api/app/controllers/spree/api/stock_items_controller.rb [hunk 13]
```diff
@@ -70,6 +61,15 @@ def scope
       def stock_item_params
         params.require(:stock_item).permit(permitted_stock_item_attributes)
       end
+
+      def count_on_hand_adjustment
+        params[:stock_item].has_key?(:count_on_hand) ? params[:stock_item][:count_on_hand].to_i : 0
```
Comment (gmacdougall) [inline api/app/controllers/spree/api/stock_items_controller.rb]: Since `nil.to_i` is `0` this can be simplified to

```
params[:stock_item][:count_on_hand].to_i
```

File: api/app/controllers/spree/api/stock_items_controller.rb [hunk 14]
```diff
@@ -70,6 +61,15 @@ def scope
...
+      end
+
+      def adjust_stock_item_count_on_hand(count_on_hand_adjustment)
+        stock_movement = @stock_location.move(@stock_item.variant, count_on_hand_adjustment, current_api_user)
+        @stock_item = stock_movement.stock_item
```
Comment (gmacdougall) [inline api/app/controllers/spree/api/stock_items_controller.rb]: Why are is `@stock_item` being re-assigned here. It's a bit of a surprise to the developer.

File: api/app/controllers/spree/api/stock_items_controller.rb (hunk 12 above)
Comment (richardnuno) [inline api/app/controllers/spree/api/stock_items_controller.rb]: They should indeed, same goes for the update. Will wrap them both in a transaction in an upcoming commit.

File: api/app/controllers/spree/api/stock_items_controller.rb (hunk 13 above)
Comment (richardnuno) [inline api/app/controllers/spree/api/stock_items_controller.rb]: Sounds good

File: api/app/controllers/spree/api/stock_items_controller.rb (hunk 14 above)
Comment (richardnuno) [inline api/app/controllers/spree/api/stock_items_controller.rb]: `@stock_item` is being reassigned because the `move` method will update the stock_item's `count_on_hand` if the variant is set to track inventory. Since `move` does a find or create of the stock item based on the variant, the change in value won't be reflected on `@stock_item`

**Comment by richardnuno** [comment]
@gmacdougall the last two commits address your suggestions. In regards to the `@stock_item` variable would a comment to indicate why the reassignment is happening be enough for this case or were you thinking of something different?

File: api/app/controllers/spree/api/stock_items_controller.rb (hunk 14 above)
Comment (gmacdougall) [inline api/app/controllers/spree/api/stock_items_controller.rb]: No, this is fine. I was just trying to understand what was going on.

Comment (richardnuno) [inline api/app/controllers/spree/api/stock_items_controller.rb]: @gmacdougall great, do you think this is mergeable with the most recent changes?

**Comment by gmacdougall** [comment]
My biggest concern remaining is that the JS doesn't have a lot of structure and is difficult to follow.

Would it be possible to organize some classes to represent different responsibilities?

**Comment by richardnuno** [comment]
Sure, will add that in an upcoming commit

**Comment by richardnuno** [comment]
@athal7 @gmacdougall Per Gregor's suggestion, I've refactored the javascript a bit. Mind taking a look?

**Comment by athal7** [comment]
i like the js structure you've put into place

File: api/app/controllers/spree/api/stock_items_controller.rb [hunk 15]
```diff
@@ -70,6 +61,35 @@ def scope
...
       end
+
+      def create_stock_item
+        success = false
+        Spree::StockItem.transaction do
+          @stock_item.save!
+          adjust_stock_item_count_on_hand(count_on_hand_adjustment)
+          success = true
+        end
+        success
```
Comment (jhawthorn) [inline api/app/controllers/spree/api/stock_items_controller.rb]: I don't believe the `success` variable has a purpose. either this method will raise an exception, or it will return true.

File: api/app/controllers/spree/api/stock_items_controller.rb [hunk 16]
```diff
@@ -70,6 +61,35 @@ def scope
...
+      end
+
+      def update_stock_item(adjustment)
+        success = false
+        Spree::StockItem.transaction do
+          @stock_item.update_attributes!(stock_item_params)
+          adjust_stock_item_count_on_hand(adjustment)
+          success = true
+        end
+        success
```
Comment (jhawthorn) [inline api/app/controllers/spree/api/stock_items_controller.rb]: Same as above.

**Comment by richardnuno** [comment]
@jhawthorn good point on the transactions, refactored them a bit in the latest commit

**Comment by magnusvk** [comment]
@cbrunsdon Wanted to follow up here. Have you had a chance to clone this down and poke at it? As it stands, I believe the code looks great and this works as intended, so it should be ready to merge. We need to get this out on our end.

**Comment by cbrunsdon** [comment]
I pulled this down and ran it and nothing came up that Richard didn't fix right away. I poked the specs and performance a bit and that is now merged as well.

The specs suffer from the fact we can't currently create a stock_item in the factory, but this isn't a good time to fix that problem. When we can we should be able to simplify the logic in the controller specs greatly.

The admin JS is an interesting bit of code. I don't have any complaints about it, just that it is now "yet another way" of how javascript in the admin is done and as we build more complex interfaces, we're going to want to standardize on how lots of it are done.

Long story short :+1: from me. Its a great feature and implemented well, Definitely better than the csv upload we generally throw in our sites.

**Comment by magnusvk** [comment]
Thanks @cbrunsdon, great feedback. Excited to ship this.
